### PR TITLE
Interface

### PR DESCRIPTION
**Pull request by yumuzi234**

### CODE REVIEW

File: pl/ast/general.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package ast
```
Comment (h8liu) [inline pl/ast/general.go]: how about `common.go`, or `base.go`

File: pl/ast/interface.go [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
-
-// Decl is a general declaration,
-// could be var, function, struct or interface
-type Decl interface{}
+// Interface declares a interface
+type Interface struct {
+	Kw       *lexing.Token
+	Name     *lexing.Token
+	Lbrace   *lexing.Token
+	Funcs    []*lexing.Token
```
Comment (h8liu) [inline pl/ast/interface.go]: these `Funcs` are actually function names. please create a new struct `InterfaceFunc` to put the names along with the sigs. don't do parallel arrays/slices.

File: pl/parse/file_test.go [hunk 3]
```diff
@@ -32,6 +32,17 @@ func TestFile_good(t *testing.T) {
...
+			add() int
+			print() string
+		}`,
+		`interface T {
+			add(int, int) int
+			print(string) 
+		}`,
+		`interface T {
+			divide(a, b int) (int, int)
+		}`,
```
Comment (h8liu) [inline pl/parse/file_test.go]: please also add a test case with named return like `f() (a int, b int)`

File: pl/parse/interface.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		ret = &ast.Interface{
+			Kw:     p.ExpectKeyword("interface"),
+			Name:   p.Expect(Ident),
+			Lbrace: p.ExpectOp("{"),
+		}
+	} else {
+		panic("expect keyword")
+	}
+	for !p.SeeOp("}") && !p.See(lexing.EOF) {
+
```
Comment (h8liu) [inline pl/parse/interface.go]: remove this blank line.

File: pl/parse/interface.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+func parseInterface(p *parser) *ast.Interface {
+	var ret *ast.Interface
+	if p.SeeKeyword("interface") {
+		ret = &ast.Interface{
+			Kw:     p.ExpectKeyword("interface"),
+			Name:   p.Expect(Ident),
+			Lbrace: p.ExpectOp("{"),
+		}
+	} else {
```
Comment (h8liu) [inline pl/parse/interface.go]: move the panic up to the beginning as `if !p.SeeKeyword()`

File: pl/parse/interface.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"shanhu.io/smlvm/lexing"
+	"shanhu.io/smlvm/pl/ast"
+)
+
+func parseInterface(p *parser) *ast.Interface {
+	var ret *ast.Interface
+	if p.SeeKeyword("interface") {
+		ret = &ast.Interface{
+			Kw:     p.ExpectKeyword("interface"),
```
Comment (h8liu) [inline pl/parse/interface.go]: you can just use `p.Shift()`

File: pl/parse/file_test.go (hunk 3 above)
Comment (h8liu) [inline pl/parse/file_test.go]: also, let's also test an empty interface.

File: pl/parse/parser.go [hunk 7]
```diff
@@ -17,7 +17,9 @@ type parser struct {
 	typeFunc func(p *parser) ast.Expr
 	stmtFunc func(p *parser) ast.Stmt
 
-	golike       bool
+	golike bool
+
+	// inlineMethod no longer supported?
```
Comment (h8liu) [inline pl/parse/parser.go]: change to `// TODO(h8liu): remove this`

File: pl/parse/struct.go [hunk 8]
```diff
@@ -68,6 +68,7 @@ func parseStruct(p *parser) *ast.Struct {
 		ret.Fields = append(ret.Fields, field)
 	}
 
+	// remove this?
```
Comment (h8liu) [inline pl/parse/struct.go]: change to `// TODO(h8liu): remove this`

File: pl/parse/interface.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"shanhu.io/smlvm/lexing"
+	"shanhu.io/smlvm/pl/ast"
+)
+
+func parseInterface(p *parser) *ast.Interface {
+	var ret *ast.Interface
+	if !p.SeeKeyword("interface") {
+		panic("expect keyword")
+	} else {
```
Comment (h8liu) [inline pl/parse/interface.go]: no need to else here
